File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque

a, b = map(int, input().split())
arr = [list(map(int, input())) for _ in range(a)]

dx, dy = [1, -1, 0, 0], [0, 0, 1, -1]

def bfs(x, y):
    q = deque()
    q.append((x, y))

    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx >= a or nx < 0 or ny >= b or ny < 0:
                continue

            if arr[nx][ny] == 0:
                continue

            if arr[nx][ny] == 1:
                arr[nx][ny] = arr[x][y] + 1
                q.append((nx, ny))

    return arr[a-1][b-1]


print(bfs(0, 0))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Queue;

public class Main {
	static int N, M;
	static int[][] arr;
	static boolean[][] visited;
	static int[] dx = { 1, -1, 0, 0 };
	static int[] dy = { 0, 0, -1, 1 };
	static boolean exit;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] st = br.readLine().split(" ");
		
		N = Integer.parseInt(st[0]);
		M = Integer.parseInt(st[1]);
		visited = new boolean[N][M];
		exit = false;
		
		arr = new int[N][M];
		
		for(int i = 0; i < N; i++) {
			st = br.readLine().split("");
			for(int j = 0; j < M; j++) {
				arr[i][j] = Integer.parseInt(st[j]);
			}
		}
		
		bfs(0, 0);
		System.out.println(arr[N-1][M-1]);
	}
	
	private static void bfs(int x, int y) {
		Queue<int[]> q = new ArrayDeque<>();
		q.offer(new int[] {x, y});
		visited[x][y] = true;
		
		while(!q.isEmpty()) {
			int[] now = q.poll();
			
			for (int i = 0; i < 4; i ++) {
				int nx = now[0] + dx[i];
				int ny = now[1] + dy[i];
				
				if (nx < 0 || nx >= N || ny < 0 || ny >= M) {
					continue;
				} 
				
				if (nx == N && ny == M) {
					exit = true;
					break;
				}
				
				if (arr[nx][ny] != 0 && !visited[nx][ny]) {
					visited[nx][ny] = true;
					arr[nx][ny] = arr[now[0]][now[1]] + 1;
					q.offer(new int[] {nx, ny});
				}
			}
			if (exit) {
				break;				
			}
		}
	}
}